File: HashQuadraticProbing.java
import java.util.Random;
import java.util.HashSet;
import java.util.Scanner;

class QuadraticProbingHashTable {
    private Integer[] table;
    private int size;
    private final int EMPTY = -1;

    public QuadraticProbingHashTable(int size) {
        this.size = size;
        table = new Integer[size];
        for (int i = 0; i < size; i++) table[i] = EMPTY;
    }

    private int hash(int key) {
        return Math.abs(key) % size;
    }

    public boolean insert(int key) {
        int index = hash(key);
        int start = index;
        int i = 1;
        while (table[index] != EMPTY && table[index] != key) {
            index = (start + i*i) % size;
            i++;
            if (i > size) return false; // full cycle
        }
        if (table[index] == key) return false;
        table[index] = key;
        return true;
    }

    public boolean delete(int key) {
        int index = hash(key);
        int start = index;
        int i = 1;
        while (table[index] != EMPTY) {
            if (table[index] == key) {
                table[index] = EMPTY;
                return true;
            }
            index = (start + i*i) % size;
            i++;
            if (i > size) break;
        }
        return false;
    }

    public boolean search(int key) {
        int index = hash(key);
        int start = index;
        int i = 1;
        while (table[index] != EMPTY) {
            if (table[index] == key) return true;
            index = (start + i*i) % size;
            i++;
            if (i > size) break;
        }
        return false;
    }

    public SearchResult searchWithDetail(int key) {
        int index = hash(key);
        int start = index;
        int i = 1;
        System.out.println("\n=== PROSES PENCARIAN QUADRATIC PROBING ===");
        System.out.println("Data dicari: " + key);
        System.out.println("Hash awal: " + index);
        System.out.println("Probing step: 1^2, 2^2, 3^2, ...");
        int step = 0;
        while (table[index] != EMPTY && step < size) {
            System.out.println("  Step " + step + ": index " + index + " -> " + table[index]);
            if (table[index] == key) {
                System.out.println("  >>> DITEMUKAN pada indeks " + index);
                return new SearchResult(true, index, step);
            }
            step++;
            index = (start + step*step) % size;
        }
        System.out.println("  >>> TIDAK DITEMUKAN setelah " + step + " langkah");
        return new SearchResult(false, -1, -1);
    }

    public void display() {
        System.out.println("\nHash Table (Quadratic Probing):");
        for (int i = 0; i < size; i++) {
            System.out.printf("%3d: ", i);
            if (table[i] == EMPTY) System.out.println("null");
            else System.out.println(table[i]);
        }
    }
}

public class HashQuadraticProbing {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        QuadraticProbingHashTable ht = new QuadraticProbingHashTable(101);
        Random rand = new Random();
        HashSet<Integer> unique = new HashSet<>();
        while (unique.size() < 100) unique.add(rand.nextInt(1000));
        for (int num : unique) ht.insert(num);
        System.out.println("100 unique random numbers inserted.");

        while (true) {
            System.out.println("\n--- MENU QUADRATIC PROBING ---");
            System.out.println("1. INPUT DATA");
            System.out.println("2. HAPUS DATA");
            System.out.println("3. CARI DATA (detail)");
            System.out.println("4. TAMPILKAN DATA");
            System.out.println("5. EXIT");
            System.out.print("Pilihan: ");
            int choice = scanner.nextInt();
            switch (choice) {
                case 1:
                    System.out.print("Angka: ");
                    int in = scanner.nextInt();
                    if (ht.search(in)) System.out.println("Sudah ada.");
                    else if (ht.insert(in)) System.out.println("Berhasil ditambahkan.");
                    else System.out.println("Tabel penuh.");
                    break;
                case 2:
                    System.out.print("Angka: ");
                    int del = scanner.nextInt();
                    if (ht.delete(del)) System.out.println("Terhapus.");
                    else System.out.println("Tidak ditemukan.");
                    break;
                case 3:
                    System.out.print("Angka: ");
                    int src = scanner.nextInt();
                    ht.searchWithDetail(src);
                    break;
                case 4:
                    ht.display();
                    break;
                case 5:
                    System.out.println("Keluar.");
                    scanner.close();
                    return;
                default: System.out.println("Pilihan salah.");
            }
        }
    }
}

File: HashQuadraticProbing.py
class QuadraticProbingHashTable:
    EMPTY = -1
    def __init__(self, size):
        self.size = size
        self.table = [self.EMPTY] * size

    def _hash(self, key):
        return abs(key) % self.size

    def insert(self, key):
        index = self._hash(key)
        start = index
        i = 1
        while self.table[index] != self.EMPTY and self.table[index] != key:
            index = (start + i*i) % self.size
            i += 1
            if i > self.size:
                return False
        if self.table[index] == key:
            return False
        self.table[index] = key
        return True

    def delete(self, key):
        index = self._hash(key)
        start = index
        i = 1
        while self.table[index] != self.EMPTY:
            if self.table[index] == key:
                self.table[index] = self.EMPTY
                return True
            index = (start + i*i) % self.size
            i += 1
            if i > self.size:
                break
        return False

    def search(self, key):
        index = self._hash(key)
        start = index
        i = 1
        while self.table[index] != self.EMPTY:
            if self.table[index] == key:
                return True
            index = (start + i*i) % self.size
            i += 1
            if i > self.size:
                break
        return False

    def search_with_detail(self, key):
        index = self._hash(key)
        start = index
        i = 1
        print("\n=== PROSES PENCARIAN QUADRATIC PROBING ===")
        print(f"Data dicari: {key}")
        print(f"Hash awal: {index}")
        print("Probing step: 1^2, 2^2, 3^2, ...")
        step = 0
        idx = index
        while self.table[idx] != self.EMPTY and step < self.size:
            print(f"  Step {step}: index {idx} -> {self.table[idx]}")
            if self.table[idx] == key:
                print(f"  >>> DITEMUKAN pada indeks {idx}")
                return True, idx, step
            step += 1
            idx = (start + step*step) % self.size
        print(f"  >>> TIDAK DITEMUKAN setelah {step} langkah")
        return False, -1, -1

    def display(self):
        print("\nHash Table (Quadratic Probing):")
        for i, val in enumerate(self.table):
            print(f"{i:3}: {'null' if val == self.EMPTY else val}")

def main():
    import random
    ht = QuadraticProbingHashTable(101)
    unique = set()
    while len(unique) < 100:
        unique.add(random.randint(0, 999))
    for num in unique:
        ht.insert(num)
    print("100 unique random numbers inserted.")

    while True:
        print("\n--- MENU QUADRATIC PROBING ---")
        print("1. INPUT DATA")
        print("2. HAPUS DATA")
        print("3. CARI DATA (detail)")
        print("4. TAMPILKAN DATA")
        print("5. EXIT")
        choice = input("Pilihan: ")
        if choice == '1':
            val = int(input("Angka: "))
            if ht.search(val):
                print("Sudah ada.")
            elif ht.insert(val):
                print("Berhasil ditambahkan.")
            else:
                print("Tabel penuh.")
        elif choice == '2':
            val = int(input("Angka: "))
            if ht.delete(val):
                print("Terhapus.")
            else:
                print("Tidak ditemukan.")
        elif choice == '3':
            val = int(input("Angka: "))
            ht.search_with_detail(val)
        elif choice == '4':
            ht.display()
        elif choice == '5':
            print("Keluar.")
            break
        else:
            print("Pilihan salah.")

if __name__ == "__main__":
    main()